### DISCUSSION TITLE
Use go:embed embed.FS rather than tar + exporting files to a temp directory?

### DISCUSSION THREAD

**Discussion by nickchomey** [php/frankenphp#1178] [Ideas]
I've made this a discussion rather than issue, as I'm not sure if it is valid or not. 

I was just exploring how FrankenPHP does the embedded builds, and it looks like
1. you set an env variable with the path of your application then run the `build-static.sh` script which creates a tar of the application which gets embedded
2. when the binary is run, the tar gets extracted to a temp directory 

I wonder if the whole process could be simplified if you use the [embed.FS](https://pkg.go.dev/embed#FS) which "implements fs.FS, so it can be used with any package that understands file system interfaces, including net/http, text/template, and html/template."? 

Moreover, it might even be higher performance since the files would be read from within the binary rather than the temp directory. Furthermore, I could see this being useful in the sense of wanting to protect Intellectual Property by not extracting php files. I dont know if this would result in extra/excessive ram usage though, as compared to the tar extraction.

I also wonder if it could perhaps even somehow be pre-compiled into the opcache?

**Comment by withinboredom** [comment]
This would need support in PHP itself (particularly the require/include/require_once/include_once directives) as well as a custom stream filter to intercept local files -- it would be similar to the `phar` implementation, just with go-specific access.

**Reply by nickchomey** [reply]
ah ok, i didnt think of how php itself needs to use the filesystem. Surely not worth the effort to implement something like this. Thanks!

**Comment by Complexicon** [comment]
Hi, i just stumbled across this whilst checking if FrankenPHP supports the Caddy builtin filesystem directives: it seems like not currenty based on the response.
Hypothetically if i would implement something like `win32/ioutil.c` (in php source) as a patch but for a go backend would that support the bulk of file io operations right? or do i need to patch other places too?

if this work one could use such cursed caddy filesystem backends as https://github.com/mohammed90/caddy-git-fs for serving php
